<a href="https://colab.research.google.com/github/ManuelaUsuga/Analitica-de-Negocios/blob/main/Reto_1_USUGA_Manuela.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Reto #1 - Manuela Usuga
**Modelo Naive Bayes**
0. Se procede con la carga de las librerias de trabajo

In [ ]:
import numpy as np #Funciones numéricas transformación y análisis de datos
import pandas as pd #Conectar el sistema con los datos

1. Se procede con la carga de la base de datos

In [ ]:
Datos='/content/1. BD1_Longitudinal Survey_Int.xlsx' #Se le da nombre a la base de datos.
Hoja=pd.read_excel(Datos,sheet_name=0) #Para sacar una hoja específica

#Se cargan las variables de interés
Hoja=Hoja[['Age', 'Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income', 'Urban']] #Recuerda poner tambien la variable de referencia

1.1 Importante limpiar los espacios Nan 💡

In [ ]:
Hoja=Hoja.dropna()

In [ ]:
#Separamos las variables por Entrada y Salida
XD=Hoja[['Age', 'Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']]
yd=Hoja[['Urban']] #Variable de referencia

2. Implementamos el modelo Naive Bayes

In [ ]:
#P1 - Para sacar el total de los datos limpios
ndatos = len(Hoja)
print('El total de los datos es', ndatos)

#Para traer la biblioteca
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd)
u2=mnb.theta_
var2=mnb.var_

std2=np.sqrt(var2)
print('La media es\n', u2)
print('La desviación estándar es\n', std2)
print('El limite superior es:\n',u2+std2)
print('El limite inferior es:\n',u2-std2)

ct = mnb.class_count_
print("Como están separados los datos", ct)
print("El total de datos es:", len(XD))
print("Los datos porcentualmente están divididos:", ct/len(XD))

#Para evaluar un solo solicitante
XS=[[34,12,15,2,24,69,160,48000]]
mnb.predict(XS) #El 1 significa



El total de los datos es 5948
La media es
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La desviación estándar es
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
El limite superior es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.12309482e+00
  2.67720628e+01 7.11135895e+01 1.74779433e+02 6.93634044e+04]]
El limite inferior es:
 [[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  2.18703841e+00
   1.77786493e+01  6.31606515e+01  1.18654593e+02 -1.52335198e+

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


array([1.])

**Analisis de resultados Naive Bayes**

De acuerdo con los resultados arrojados por el modelo, el número total de los datos es de 5948. Los datos porcentualmente están divididos en 20.83% para las personas que viven en zonas no urbanas, y del 79.16% para aquellas personas que residen en zonas urbanas.

Para las variables, la Edad tiene un intervalo de entre limite superior 20 años y limite inferior de 15 años para aquellos que viven en area rural y urbana.

la Altura tiene un intervalo de entre limite superior 71 y limite inferior 63 pulgadas para aquellos que viven en area rural y limite superior 71 y limite inferior 62 pulgadas para area urbana.

la Peso tiene un intervalo de entre limite superior 179 y limite inferior 118 libras para area rural. Adicionalmente, de limite inferior 174 y limite superior 114 libras para aquellos que viven en area urbana.

Una persona con 34 años de edad, con una madre con 12 años de eduacion y un padre con 15 años de educación, con 2 miembros de familia, un puntaje de autoestima de 69 y una altura de 160 pulgadas y un ingreso de 48,000; es un individuo que reside en area urbana.

In [ ]:
import pandas as pd

# Obtener las predicciones del modelo
ydp = mnb.predict(XD)

# Crear un DataFrame con las predicciones y la variable 'Weight'
predicciones_df = pd.DataFrame({'Urban_pred': ydp, 'Weight': Hoja['Weight']})

# Filtrar los datos para la categoría 0 de 'Urban_pred'
categoria_0_pred = predicciones_df[predicciones_df['Urban_pred'] == 0]

# Obtener el valor mínimo de 'Weight' para la categoría 0 predicha
valor_minimo_pred = categoria_0_pred['Weight'].min()

# Imprimir el valor mínimo
print("El valor mínimo de 'Weight' para la categoría 0 (predicciones) es:", valor_minimo_pred)

El valor mínimo de 'Weight' para la categoría 0 (predicciones) es: 140.0


**Matriz de confusión**

Se procede con la evaluación (Desempeño del modelo)


In [ ]:
ydp=mnb.predict(XD) #Hazme el pronostico de los positivos y negativos de Urban
from sklearn.metrics import confusion_matrix #Para sacar de biblioteca
cm=confusion_matrix(yd,ydp)
print(cm) #VP, FP, VN, VP

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

#Exactitud = VN + VP / Total de datos #SIEMPRE
Ex=(VN+VP)/ (VN+VP+FN+FP) #Desempeño general
print("La exactitud alcanzada por el modelo es" , Ex)

#Tasa de error = FN + FP / Total de los datos
Te=(FN+FP) / (VN+VP+FN+FP)
print("La tasa de error es" , Te)

#Sensibilidad
Sen = VP/(VP+FN)
print("La sensibilidad del modelo es de:", Sen)

#Especificidad = (VN)/(FP+VN)
Especificidad = (VN)/(FP+VN)
print("La especificidad del modelo es de:", Especificidad)

#Precisión
Precision=(VP)/(VP+FP) #Comportamiento general identificando preaprobados
print("La precisión del modelo es de:", Precision)

#Precisión negativa
Precision_neg=(VN)/(VN+FN)
print("La precisión negativa del modelo es de:", Precision_neg)

[[   3 1236]
 [   8 4701]]
La exactitud alcanzada por el modelo es 0.7908540685944856
La tasa de error es 0.20914593140551446
La sensibilidad del modelo es de: 0.9983011255043533
La especificidad del modelo es de: 0.002421307506053269
La precisión del modelo es de: 0.7918140474987367
La precisión negativa del modelo es de: 0.2727272727272727


El modelo alcanzó una exactitud del 79.09 %, lo que indica que, en general, clasifica correctamente la mayoría de los casos. Sin embargo, la tasa de error del 20.91 % sugiere que todavía hay una proporción significativa de predicciones incorrectas. La sensibilidad del modelo es de 99.83 %, lo que demuestra su alta capacidad para identificar correctamente los casos positivos. No obstante, la especificidad es baja (0.24 %), lo que significa que el modelo tiene dificultades para identificar correctamente los casos negativos, generando una gran cantidad de falsos positivos. Además, la precisión del modelo es del 79.18 %, lo que implica que un porcentaje considerable de las predicciones positivas son correctas. Finalmente, la precisión negativa es de 27.27 %, lo que sugiere que la capacidad del modelo para predecir correctamente los casos negativos es muy limitada. Esto indica un desequilibrio en el desempeño del modelo, favoreciendo la identificación de los casos positivos a costa de los negativos.